lib/AstOfLlbc.ml

```diff
@@ -566,47 +680,190 @@ let maybe_addrof (env: env) (ty: C.ty) (e: K.expr) =
+      List.map (fun (item_name, decl_id) ->
+        match decl_id with
+        | Some decl_id ->
+            let decl = env.get_nth_function decl_id in
+            (clause_id, item_name), (tvarid, trait_decl.C.name, decl.C.signature)
+        | None ->
+            failwith ("TODO: handle provided trait methods, like " ^ item_name)
+            ) trait_decl.C.provided_methods
+    end
+  ) trait_clauses
+
+(* Interpret a Rust function type, with trait bounds, into the krml Ast, providing:
+   - the number of polymorphic type variables
+   - the cg types, which only contains the original Rust const generic variables (not trait methods)
+   - the argument types, prefixed by the dictionary-style passing of trait clause methods
+   - the return type
+   - whether the function is assumed, or not. *)
 type lookup_result = {
-  name: K.lident;
   n_type_args: int; (* just for a sanity check *)
   cg_types: K.typ list;
   arg_types: K.typ list;
   ret_type: K.typ;
-  is_assumed: bool;
+  is_known_builtin: bool;
 }
 
-let lookup_fun (env: env) (f: C.fn_ptr): lookup_result =
+let rec lookup_signature env depth signature =
+  let { C.generics = { types = type_params; const_generics; trait_clauses; _ }; inputs; output; _ } = signature in
+  L.log "Calls" "%s--> args: %s, ret: %s" depth
+    (String.concat " ++ " (List.map (Charon.PrintTypes.ty_to_string env.format_env) inputs))
+    (Charon.PrintTypes.ty_to_string env.format_env output);
+  let env = push_cg_binders env const_generics in
+  let env = push_type_binders env type_params in
+
+  let clause_mapping = build_trait_clause_mapping env trait_clauses in
+  debug_trait_clause_mapping env clause_mapping;
+  let clause_binders = mk_clause_binders_and_args env clause_mapping in
+  let clause_ts = List.map (fun { t; _ } -> t) clause_binders in
+
+  {
+    n_type_args = List.length type_params;
+    cg_types = List.map (fun (v: C.const_generic_var) -> typ_of_literal_ty env v.ty) const_generics;
+    arg_types = clause_ts @ List.map (typ_of_ty env) inputs;
+    ret_type = typ_of_ty env output;
+    is_known_builtin = false
+  }
+
+(* Assumes type variables have been suitably bound in the environment *)
+and mk_clause_binders_and_args env clause_mapping: clause_binder list =
+  List.map (fun ((clause_id, item_name), (tvarid, trait_name, (signature: C.fun_sig))) ->
+    let tvar = lookup_typ env tvarid in
+    let t = thd3 (typ_of_signature env signature) in
+    let t = Krml.DeBruijn.subst_t (TBound tvar) 0 t in
+    let pretty_name = string_of_name env trait_name ^ "_" ^ item_name in
+    let sig_info = {
+      n_type_args = List.length signature.generics.types - 1;
+      n_cg_args = List.length signature.generics.const_generics;
+    } in
+    { pretty_name; t; clause_id; item_name; sig_info }
+  ) clause_mapping
+
+
+(* Transforms a lookup result into a usable type, taking into account the fact that the internal Ast
+   is ML-style and does not have zero-argument functions. *)
```

Comment:
What do you do for the functions defining the bodies of globals (which are functions with 0 arguments)?

Reply:
Dedicated treatment (see https://github.com/AeneasVerif/eurydice/blob/main/lib/AstOfLlbc.ml#L1152-L1166) -- I directly call into expression_of_raw_statement without going through lookup_function